src/RegistryCoordinator.sol

```diff
@@ -647,19 +658,31 @@ contract RegistryCoordinator is
         );
         require(
             quorumsToRemove.isSubsetOf(currentBitmap),
-            "RegistryCoordinator._deregisterOperator: operator is not registered for specified quorums"
+            "RegistryCoordinator._deregisterOperator: not registered for quorum"
         );
         uint192 newBitmap = uint192(currentBitmap.minus(quorumsToRemove));
 
         // Update operator's bitmap and status
         _updateOperatorBitmap({operatorId: operatorId, newBitmap: newBitmap});
 
-        // If the operator is no longer registered for any quorums, update their status and deregister
-        // them from the AVS via the EigenLayer core contracts
-        if (newBitmap.isEmpty()) {
-            operatorInfo.status = OperatorStatus.DEREGISTERED;
-            serviceManager.deregisterOperatorFromAVS(operator);
-            emit OperatorDeregistered(operator, operatorId);
+
+        bool operatorSetAVS = IAVSDirectory(serviceManager.avsDirectory()).isOperatorSetAVS(address(serviceManager)); /// TODO: call avsdirectory
+        if (operatorSetAVS){
+            bytes memory quorumBytes = BitmapUtils.bitmapToBytesArray(quorumsToRemove);
+            uint32[] memory operatorSetIds = new uint32[](quorumBytes.length);
+            for (uint256 i = 0; i < quorumBytes.length; i++) {
+                operatorSetIds[i] = uint8(quorumBytes[i]);
+            }
+            serviceManager.deregisterOperatorFromOperatorSets(operator, operatorSetIds);
+
+        } else {
+            // If the operator is no longer registered for any quorums, update their status and deregister
+            // them from the AVS via the EigenLayer core contracts
+            if (newBitmap.isEmpty()) {
+                operatorInfo.status = OperatorStatus.DEREGISTERED;
+                serviceManager.deregisterOperatorFromAVS(operator);
+                emit OperatorDeregistered(operator, operatorId);
```

Comment:
We still want to emit this event even in operatorSet-AVS case?

Reply:
i think so. I was wondering if there were any offchain dependencies on this event. it seemed reasonable to keep. This would be if they were removed deregistered from all operators sets they were a part of

Reply:
I see what you mean. Moved the OperatorRegistered event for M2 to only be emitted if they're still an M2 AVS [6a77df8](https://github.com/Layr-Labs/eigenlayer-middleware/pull/301/commits/6a77df8cf257308f4df87bde96d692ede2eff3ce)